fix #17419 useOneOfInterfaces option for java clients & servers

Nicklas2751 opened the pull request:
Some java generators generated wrong, not compiling code, when using the option `useOneOfInterfaces`.
The bug was caused by old/wrong `model.mustache`files without the part supporting the `useOneOfInterfaces` option.
These clients were affected:
- jersey2
- jersey3
- microprofile
- native
- okhttp-gson 

as well as the following servers: 
- helidon
- jaxrs
- mirconaut
- microprofile
- play
- pkmst

This PR fixes all of them 😎

It also fixes a compile error of the java native client generator when using anyOf with generic types.

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package || exit
  ./bin/generate-samples.sh ./bin/configs/*.yaml || exit
  ./bin/utils/export_docs_generators.sh || exit
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming `7.x.0` minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@bbdouglas (2017/07) @sreeshas (2017/08) @jfiala (2017/08) @lukoyanov (2017/09) @cbornet (2017/09) @jeff9finger (2018/01) @karismann (2019/03) @Zomzog (2019/04) @lwlee2608 (2019/10) @martin-mfg (2023/08)

Nicklas2751 committed:
fix (JAVA NATIVE CLIENT): using anyOf with generic type lead to compile error

Nicklas2751 committed:
fix (JAVA CLIENTS): fix useOneOfInterfaces not generating interfaces for java clients jersey2, jersey3, microprofile, native & okhttp-gson (#17419)

Nicklas2751 committed:
fix (JAVA HELIDON JAXRS MICRONAUT MICROPROFILE PLAY PKMST): fix useOneOfInterfaces not generating interfaces for java server helidon, jaxrs, micronaut, microprofile, play and pkmst (#17419)

Nicklas2751 committed:
refactor: remove unused import

Nicklas2751 committed:
fix: okhttp-gson model indentation

Nicklas2751 committed:
refactor (JAVA HELIDON SE SERVER): remove empty line

Nicklas2751 committed:
refactor (JAVA JAX-RS SERVER): remove empty line

Nicklas2751 commented:
@wing328 Seems like the others don't have any feedback on this 😅

martin-mfg reviewed (changes requested):
Hi @Nicklas2751, thanks for the PR! I found 2 problems:
- When I run OpenAPI Generator with generator=java, library=okhttp-gson, additionalProperties:useOneOfInterfaces=true on [this input spec](https://github.com/OpenAPITools/openapi-generator/blob/fef84d956e23dad2cb6ef49b0519b7427a13857a/modules/openapi-generator/src/test/resources/3_0/typescript-fetch/oneOf.yaml), the output fails to compile because `org.openapitools.client.model.TestResponse.CustomTypeAdapterFactory` doesn't exist.
- Using library=jersey2 with all other settings like above generates a file docs/TestResponse.md with sample code which doesn't compile.

I guess both problems apply to more than just the one mentioned library.

Could you please fix these problems? Thanks!

<sub>(I did not review the whole PR. So when the described problems are fixed, further review is necessary.)</sub>

Nicklas2751 commented:
@martin-mfg `org.openapitools.client.model.TestResponse.CustomTypeAdapterFactory` shouldn't exists and `org.openapitools.client.model.TestResponse` should be an interface. 

I added another test to `JavaClientCodegenTest` with the OpenAPI schema you named and it worked. But when I use the CLI it doesn't work like you said. I added some more test but couldn't recreate the problem. But then I saw that I missed to regenerate the CLI Jar so I cleared the target folder and ran `./mvnw clean package` again. Now the CLI works correctly too.

This is how I started the CLI:
```
java -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate \
  -i https://github.com/OpenAPITools/openapi-generator/raw/fef84d956e23dad2cb6ef49b0519b7427a13857a/modules/openapi-generator/src/test/resources/3_0/typescript-fetch/oneOf.yaml \
  -g java \
  --library okhttp-gson \
  --additional-properties useOneOfInterfaces=true \
  -o test/okhttp-gson
```


tl;dr I had to regenerate the CLI then it generated the Interface as expected and without any compile errors.

Nicklas2751 committed:
test (JAVA CLIENTS): add another test for the useOneOfInterfaces option

Nicklas2751 committed:
fix (JAVA CLIENTS): don't generate oneOf (not working) example code when using useOneOfInterfaces option

Nicklas2751 commented:
@martin-mfg The second problem was in the doc generation of jersey2, jersey3 and native. I fixed it & added an test for it :)

martin-mfg commented:
> tl;dr I had to regenerate the CLI then it generated the Interface as expected and without any compile errors.

It seems we get different results, or maybe we're talking about different things. So I extended a unit test to show the problem I get - please see https://github.com/martin-mfg/openapi-generator/pull/77/commits/554bb00927331a2e30520917a1bece31b9d93a2b. The test fails, as seen in [this pipeline](https://github.com/martin-mfg/openapi-generator/actions/runs/14149466847/job/39640838951?pr=77).

As you said, `org.openapitools.client.model.TestResponse.CustomTypeAdapterFactory` indeed doesn't exist. But still it is being referenced from `src/main/java/org/openapitools/client/JSON.java`.

Nicklas2751 committed:
fix (JAVA OKHTTP-GSON): don't add oneOf interface as CustomTypeAdapterFactory

Nicklas2751 commented:
@martin-mfg Ah, I see! Added another test for it & fixed it :)

Nicklas2751 commented:
@martin-mfg *ping*  🔔 :)

martin-mfg reviewed (commented):
Hi, Sorry for the delay. I am rather busy at the moment.

I tested the changes with the okhttp-gson library and [this input spec](https://gist.github.com/martin-mfg/4e61a166b6b1a65552ca6d2e56c7b26c). The generated output contains a few problems:
- `TestObject` implements neither `MyOperationRequest` nor `MyOperation200Response`
- It's unclear how to convert a String array to `MyOperationRequest` or how to convert `MyOperation200Response` to a number array.
- The readme contains example code including this line:
   ```java
    MyOperationRequest myOperationRequest = new MyOperationRequest(); // MyOperationRequest | 
   ```
   Since `MyOperationRequest` is an interface, `new MyOperationRequest()` doesn't work.

I guess these problems apply also to some other libraries in addition to okhttp-gson.

Could you please fix these problems? Thanks!

<sup>(I did not review the whole PR. So when the described problems are fixed, further review is necessary.)</sup>

--------------------


> @martin-mfg Is there anything I can do to make it easier for you to review this?

Don't feel obligated to get an approval from me specifically. It's totally fine if wing328 approves and merges this PR directly. Or if any other member of the Java technical committee approves this PR.

It might also be helpful if you do more testing of special cases on your own. You could look at some [existing OpenAPI files](https://github.com/OpenAPITools/openapi-generator/tree/fef84d956e23dad2cb6ef49b0519b7427a13857a/modules/openapi-generator/src/test/resources/3_0), the [OpenAPI documentation](https://swagger.io/docs/specification/v3_0/basic-structure/) or the [OpenAPI specification](https://spec.openapis.org/oas/v3.0.3.html) and try if the most relevant features work in combination with your `useOneOfInterfaces` changes.

Finally, we can also discuss a bit more via Slack.
Invitation link for account creation: https://join.slack.com/t/openapi-generator/shared_invite/zt-36ucx4ybl-jYrN6euoYn6zxXNZdldoZA
Link to my Slack profile: https://openapi-generator.slack.com/team/U051HB9A0FJ

Nicklas2751 committed:
fix (JAVA): example in README for useOneOfInterfaces (#17419)

Nicklas2751 committed:
chore (JAVA): regenerate samples and export docs

Nicklas2751 commented:
@martin-mfg I created a new issue for the useless interface which is generated for oneOf with array types: #23494

I decided to create a new issue since the main goal of this PR is to support the `useOneOfInterfaces` option for all java generators

pjurczen commented:
@martin-mfg can we get this merged?

Ignacio-Vidal commented:
I just realised I was working in parallel on the same, but only for the `java-client/microprofile` and `jaxrs-spec/quarkus` generators. I found a bug in the jaxrs-spec generator when resolving the JsonTypeInfo annotation for the child classes, where the generator doesn't resolve the value from the `mapping`. See https://github.com/OpenAPITools/openapi-generator/pull/24174 with an example spec for testing
